packages/loader/container-loader/src/container.ts

```diff
@@ -933,6 +933,42 @@ export class Container
 		);
 	}
 
+	/**
+	 * Log close and/or dispose telemetry events.
+	 * Close and Dispose are semantically very similar and we want to ensure the logging is clear
+	 * regardless of the exact scenario (close only v. close-then-dispose v. dispose)
+	 */
+	private logCloseOrDispose(path: "close" | "dispose", error?: ICriticalContainerError) {
```

Comment:
you should validate we don't log the same error multiple times.

Reply:
We will definitely log the same error multiple times.  This is fine, errors get logged repeatedly all the time.

Reply:
I guess you're thinking about stress tests / e2e tests where we are just looking at raw error counts.  I'll think about it for a sec...

Reply:
the changes being made are a fundamental change to our shutdown behavior, it think to some degree we will need to adapt our querying to deal with it.

Reply:
> you should validate we don't log the same error multiple times.

ContainerDispose is `generic` event so we're fine.

> to some degree we will need to adapt our querying to deal with it.

I think the detail of dispose v. close is mostly not interesting to us.  In my mind dispose is just close plus some necessary teardown.  In terms of session lifecycle, it's going to be a whole lot simpler for everyone to keep the mental model of "close event ends the session".

Reply:
i also have no idea what the expectation are between close and dispose state, like should we even log anything after close? if dispose closes, just logging close at dispose and shutting down the logger might work. I don't think that is what we want, but i really don't know.

Reply:
Here's my take:

* From FF's session lifecycle standpoint, all we care about is Close.
* And being able to easily monitor Close is important.  So whether someone calls `close` or `dispose`, we should get the `ContainerClose` event immediately (and it indicates which was called via `Data_details`)
* I don't know when we'd care to see a log for Dispose-After-Close, but probably good to have it.

Hence, this change.  After some thought I'd update my PR slightly to do `else if (disposing)` and change that event name to `ContainerDisposeAlreadyClosed` or something.

I'll share this PR / thread with some others to get more feedback before proceeding.

Reply:
Close is to be called when a critical error occurs and we want to switch the container to a safe, read-only state. Because the document service and runtime are no longer disposed when the container is closed, we can, for example, delay-load data stores / DDSs (as user scrolls through UX) but can't, for example, submit a summary. As with normal flows, once the container is no longer expected to be used, it should be disposed.

Dispose is to be called when the container is no longer expected to be used (i.e. closing the page).

I still think `Container.close(...)` should log the "ContainerClose" error as that telemetry is useful (to know when critical errors occur). I don't think `Container.dispose(...)` should log this event as they are technically two separate things now.

Reply:
Great example of what's different between them.  But we all agree `dispose` is a superset of `close`, so I still like having a chokepoint.

I think we still have something to learn about host behavior of close-then-later-dispose, btw

Reply:
>> But we all agree dispose is a superset of close, so I still like having a chokepoint.

is that true, i'm under the impression that we do things during close that we don't do during dispose, so the view from the container author is different, and could lead to different side effects, so i don't think we can view it that way either.

Reply:
basically, closing and disposing are separate states. Frequently we will close then dispose, but not always, and directly disposing is not the same as closing.

Reply:
They are somewhat super-sets, but it's likely not a correct/good way to think about it. From outside perspective, `dispose` is cleaning up all resources of a container and not touching it again. And `close` is "Hey, I hit this critical error and need to switch to read-only so nothing bad happens".